File: problems/ruby/090721-minimum-depth-of-binary-tree/code.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end

# https://leetcode.com/problems/minimum-depth-of-binary-tree/submissions/
#
# @param {TreeNode} root
# @return {Integer}

def min_depth(root)
  min_depth_helper(root).infinite? ? 0 : min_depth_helper(root)
end

def min_depth_helper(root)
  return Float::INFINITY if root.nil?

  return 1 if leaf?(root)

  1 + [min_depth_helper(root.right), min_depth_helper(root.left)].min
end

def leaf?(root)
  root && (root.left || root.right).nil?
end

# shorter solution
def better_min_depth(root)
  return 0 if root.nil?

  left_height = better_min_depth(root.left)
  right_height = better_min_depth(root.right)

  if (left_height === 0 || right_height === 0)
    return 1 + left_height + right_height
  end

  1 + [left_height, right_height].min
end

File: problems/javascript/09072021-minimum-depth-of-binary-tree/code.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

const isLeaf = (node) => node && !(node.left || node.right)

const minDepthHelper = (root) => {
  if (!root) return Infinity

  if (isLeaf(root)) return 1

  return 1 + Math.min(minDepthHelper(root.left), minDepthHelper(root.right))
}

/**
 * https://leetcode.com/problems/minimum-depth-of-binary-tree/
 *
 * @param {TreeNode} root
 * @return {number}
 */
const minDepth = (root) => {
  return !isFinite(minDepthHelper(root)) ? 0 : minDepthHelper(root)
}

// shorter solution
const betterMinDepth = (root) => {
  if (!root) return 0

  const leftHeight = betterMinDepth(root.left)
  const rightHeight = betterMinDepth(root.right)

  if (leftHeight === 0 || rightHeight === 0) {
    return 1 + leftHeight + rightHeight
  }

  return 1 + Math.min(leftHeight, rightHeight)
}
